### DISCUSSION TITLE
Lesson 9: Compile error - stack too deep

### DISCUSSION THREAD

**Discussion by RichardPalladino** [Cyfrin/foundry-full-course-cu#345] [Q&A] [answered]
Lesson 9 was going great until I got to the point where we refactored the code and added "deployerKey".  There were some other modifications with this, like grabbing the private key from my .env file with `vm.envUint("SEPOLIA_PRIVATE_KEY")`, but for the most part little else changed.

Yet when I tried to build after making the changes and clearing the code errors, I received the "Compile error - stack too deep" error.

I know to use the `--via-ir` switch option, but even that took literal minutes to compile.

Any idea of the root cause, or how I can keep this from happening in the future?  I don't have any unused variables.  I'm trying to figure out better optimizations.  Stanzas don't seem too long.  I just don't get how a simple change like that may have impacted the compiler so much.

**Comment by Naman1729** [comment]
Hey @RichardPalladino , Can you send you github repo . So, we can easily debug this error

**Reply by RichardPalladino** [reply]
https://github.com/RichardPalladino/f23-smartcontract-lottery

I'm going line-by-line of the key files now, comparing with what Patrick has done.

Thank you.

**Reply by sharonphiliplima** [reply]
Hey, even I am facing the same error. Please help me if you have found any fix. 
My entire learning is halted rn.

**Comment by 0xusmanf** [comment, marked as answer]
@RichardPalladino following line in your `RaffleTest.t.sol` file is causing this error. `(entranceFee, interval, vrfCoordinator, gasHash, subscriptionId, callbackGasLimit, linkToken, deployerKey) = helperConfig.activeNetworkConfig();` try removing one variable from this tuple and your code will work fine. i.e. you can remove `subscriptionId` as you are not using it in this file. your line should look like this after removing this variable, `(entranceFee, interval, vrfCoordinator, gasHash, , callbackGasLimit, linkToken, deployerKey) = helperConfig.activeNetworkConfig();`. @sharonphiliplima I think you are facing the same issue, try this if it works. Otherwise, you can share your repo here so I can debug.

**Reply by sharonphiliplima** [reply]
```
function setUp() external {
        DeployRaffle deployer = new DeployRaffle();
        (raffle, helperConfig) = deployer.run();
        (
            entranceFee,
            interval,
            vrfCoordinator,
            gasLane,
            subscriptionId,
            callbackGasLimit,
            link,
            deployerKey
        ) = helperConfig.activeNetworkConfig(); //deconstructing basically!
        vm.deal(PLAYER, STARTING_USER_BALANCE);

        if (subscriptionId == 0) {
            CreateSubscription createSubscription = new CreateSubscription();
            subscriptionId = createSubscription.createSubscription(
                vrfCoordinator,
                deployerKey
            );
        }
```

I think that we are using `subscriptionId` in the `RaffleTest.t.sol`
so, I removed `gasLane` did I do right?

**Reply by sharonphiliplima** [reply]
Also, would it be viable to 

```
struct RaffleState {
        Raffle raffle;
        HelperConfig helperConfig;
        address player;
        uint256 entranceFee;
        uint256 interval;
        address vrfCoordinator;
        bytes32 gasLane;
        uint64 subscriptionId;
        uint32 callbackGasLimit;
        address link;
        uint256 deployerKey;
    }
```

turn the state variables into a struct in the test file?

**Reply by RichardPalladino** [reply]
@sharonphiliplima I'm not sure that second approach - using a struct - would solve any issues since what @usmanfarooq91 calls out as the issue is referencing too many variables in the tuple receiving the return values.  

It wouldn't matter if it was 
`(raffleState.entranceFee, raffleState.interval, raffleState.vrfCoordinator, raffleState.gasHash, raffleState.subscriptionId, raffleState.callbackGasLimit, raffleState.linkToken, raffleState.deployerKey) = helperConfig.activeNetworkConfig();` or 
`(entranceFee, interval, vrfCoordinator, gasHash, subscriptionId, callbackGasLimit, linkToken, deployerKey) = helperConfig.activeNetworkConfig();`

It still probably would give a stack too deep error, either way.

The only other way I've solved the issue live on a chain is by using arrays, not structs, because those reduce the number of pointers.

I just now removed my two gas variables (gasHash and callbackgasLimit), ran forge test, and this fix solved the problem for me.  Good luck, and excellent persistence.

**Reply by 0xusmanf** [reply]
@sharonphiliplima Yes you can remove `gasLane`, it doesn't matter which variable you remove as long as you remove one.

**Comment by a1111198** [comment]
I solved my issuse by changing address(raffle) from stack to Memory and not removing any of the above varibale,

from

```solidity
 addConsumer.addConsumer(
            vrfCoordinator,
            subscriptionId,
            address(raffle),
            deployerKey
        );
 ```
  To

```solidity
address raffleAddress = address(raffle);
        addConsumer.addConsumer(
            vrfCoordinator,
            subscriptionId,
            raffleAddress,
            deployerKey
        );
  ```
